[Bug]: Storybook automigration fails to install dependencies in pnpm workspace

### Describe the bug

When running `pnpx storybook@next upgrade --prerelease` in a pnpm workspace project, the migrations fail when trying to install dependencies:

``` 
🔎 found a 'eslintPlugin' migration:
╭ Automigration detected ──────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                                      │
│   We've detected you are not using our eslint-plugin.                                                                                │
│                                                                                                                                      │
│   In order to have the best experience with Storybook and follow best practices, we advise you to install eslint-plugin-storybook.   │
│                                                                                                                                      │
│   More info: https://github.com/storybookjs/eslint-plugin-storybook#readme                                                           │
│                                                                                                                                      │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
✔ Do you want to run the 'eslintPlugin' migration on your project? … yes
✅ Adding dependencies: eslint-plugin-storybook
 ERR_PNPM_ADDING_TO_ROOT  Running this command will add the dependency to the workspace root, which might not be what you want - if you really meant it, make it explicit by running this command again with the -w flag (or --workspace-root). If you don't want to see this warning anymore, you may set the ignore-workspace-root-check setting to true.

 ERROR  Command failed with exit code 1: /Users/Library/pnpm/store/v3/tmp/dlx-5676/node_modules/.bin/storybook upgrade --prerelease
```

Storybook should either detect that this is a workspace and install the dependencies with the `--workspace-root` flag, or it should prompt me to install the dependency manually

--- a/code/lib/cli/src/js-package-manager/NPMProxy.ts
+++ b/code/lib/cli/src/js-package-manager/NPMProxy.ts
@@ -6,8 +6,6 @@ export class NPMProxy extends JsPackageManager {
 
   installArgs: string[] | undefined;
 
-  uninstallArgs: string[] | undefined;
-
   initPackageJson() {
     return this.executeCommand('npm', ['init', '-y']);
   }
@@ -26,16 +24,9 @@ export class NPMProxy extends JsPackageManager {
 
   getInstallArgs(): string[] {
     if (!this.installArgs) {
-      this.installArgs = ['install'];
+      this.installArgs = [];
     }
     return this.installArgs;
-  }
-
-  getUninstallArgs(): string[] {
-    if (!this.uninstallArgs) {
-      this.uninstallArgs = ['uninstall'];
-    }
-    return this.uninstallArgs;
   }
 
   public runPackageCommand(command: string, args: string[], cwd?: string): string {
@@ -52,7 +43,7 @@ export class NPMProxy extends JsPackageManager {
   }
 
   protected runInstall(): void {
-    this.executeCommand('npm', this.getInstallArgs(), 'inherit');
+    this.executeCommand('npm', ['install', ...this.getInstallArgs()], 'inherit');
   }
 
   protected runAddDeps(dependencies: string[], installAsDevDependencies: boolean): void {
@@ -62,13 +53,13 @@ export class NPMProxy extends JsPackageManager {
       args = ['-D', ...args];
     }
 
-    this.executeCommand('npm', [...this.getInstallArgs(), ...args], 'inherit');
+    this.executeCommand('npm', ['install', ...this.getInstallArgs(), ...args], 'inherit');
   }
 
   protected runRemoveDeps(dependencies: string[]): void {
     const args = [...dependencies];
 
-    this.executeCommand('npm', [...this.getUninstallArgs(), ...args], 'inherit');
+    this.executeCommand('npm', ['uninstall', ...this.getInstallArgs(), ...args], 'inherit');
   }
 
   protected runGetVersions<T extends boolean>(

--- a/code/lib/cli/src/js-package-manager/PNPMProxy.ts
+++ b/code/lib/cli/src/js-package-manager/PNPMProxy.ts
@@ -1,3 +1,4 @@
+import { pathExistsSync } from 'fs-extra';
 import { JsPackageManager } from './JsPackageManager';
 import type { PackageJson } from './PackageJson';
 
@@ -6,7 +7,12 @@ export class PNPMProxy extends JsPackageManager {
 
   installArgs: string[] | undefined;
 
-  uninstallArgs: string[] | undefined;
+  detectWorkspaceRoot() {
+    const CWD = process.cwd();
+
+    const pnpmWorkspaceYaml = `${CWD}/pnpm-workspace.yaml`;
+    return pathExistsSync(pnpmWorkspaceYaml);
+  }
 
   initPackageJson() {
     return this.executeCommand('pnpm', ['init', '-y']);
@@ -24,6 +30,17 @@ export class PNPMProxy extends JsPackageManager {
     return this.executeCommand('pnpm', ['--version']);
   }
 
+  getInstallArgs(): string[] {
+    if (!this.installArgs) {
+      this.installArgs = [];
+
+      if (this.detectWorkspaceRoot()) {
+        this.installArgs.push('-w');
+      }
+    }
+    return this.installArgs;
+  }
+
   runPackageCommand(command: string, args: string[], cwd?: string): string {
     return this.executeCommand(`pnpm`, ['exec', command, ...args], undefined, cwd);
   }
@@ -38,7 +55,7 @@ export class PNPMProxy extends JsPackageManager {
   }
 
   protected runInstall(): void {
-    this.executeCommand('pnpm', ['install'], 'inherit');
+    this.executeCommand('pnpm', ['install', ...this.getInstallArgs()], 'inherit');
   }
 
   protected runAddDeps(dependencies: string[], installAsDevDependencies: boolean): void {
@@ -48,13 +65,13 @@ export class PNPMProxy extends JsPackageManager {
       args = ['-D', ...args];
     }
 
-    this.executeCommand('pnpm', ['add', ...args], 'inherit');
+    this.executeCommand('pnpm', ['add', ...args, ...this.getInstallArgs()], 'inherit');
   }
 
   protected runRemoveDeps(dependencies: string[]): void {
     const args = [...dependencies];
 
-    this.executeCommand('pnpm', ['remove', ...args], 'inherit');
+    this.executeCommand('pnpm', ['remove', ...args, ...this.getInstallArgs()], 'inherit');
   }
 
   protected runGetVersions<T extends boolean>(

--- a/code/lib/cli/src/js-package-manager/Yarn1Proxy.test.ts
+++ b/code/lib/cli/src/js-package-manager/Yarn1Proxy.test.ts
@@ -42,7 +42,11 @@ describe('Yarn 1 Proxy', () => {
 
       yarn1Proxy.installDependencies();
 
-      expect(executeCommandSpy).toHaveBeenCalledWith('yarn', [], expect.any(String));
+      expect(executeCommandSpy).toHaveBeenCalledWith(
+        'yarn',
+        ['install', '--ignore-workspace-root-check'],
+        expect.any(String)
+      );
     });
   });
 
@@ -69,7 +73,7 @@ describe('Yarn 1 Proxy', () => {
 
       expect(executeCommandSpy).toHaveBeenCalledWith(
         'yarn',
-        ['add', '-D', '--ignore-workspace-root-check', '@storybook/preview-api'],
+        ['add', '--ignore-workspace-root-check', '-D', '@storybook/preview-api'],
         expect.any(String)
       );
     });

--- a/code/lib/cli/src/js-package-manager/Yarn1Proxy.ts
+++ b/code/lib/cli/src/js-package-manager/Yarn1Proxy.ts
@@ -4,6 +4,36 @@ import type { PackageJson } from './PackageJson';
 export class Yarn1Proxy extends JsPackageManager {
   readonly type = 'yarn1';
 
+  installArgs: string[] | undefined;
+
+  getInstallArgs(): string[] {
+    if (!this.installArgs) {
+      this.installArgs = [];
+
+      if (this.detectWorkspaceRoot()) {
+        this.installArgs.push('-W');
+      }
+    }
+    return this.installArgs;
+  }
+
+  detectWorkspaceRoot() {
+    const { workspaces } = this.readPackageJson();
+
+    if (Array.isArray(workspaces)) {
+      return workspaces.length > 0;
+    }
+
+    if (workspaces && typeof workspaces === 'object') {
+      if (workspaces.packages) {
+        return workspaces.packages.length > 0;
+      }
+      return false;
+    }
+
+    return false;
+  }
+
   initPackageJson() {
     return this.executeCommand('yarn', ['init', '-y']);
   }
@@ -30,7 +60,7 @@ export class Yarn1Proxy extends JsPackageManager {
   }
 
   protected runInstall(): void {
-    this.executeCommand('yarn', [], 'inherit');
+    this.executeCommand('yarn', ['install', ...this.getInstallArgs()], 'inherit');
   }
 
   protected runAddDeps(dependencies: string[], installAsDevDependencies: boolean): void {
@@ -40,13 +70,13 @@ export class Yarn1Proxy extends JsPackageManager {
       args = ['-D', ...args];
     }
 
-    this.executeCommand('yarn', ['add', ...args], 'inherit');
+    this.executeCommand('yarn', ['add', ...args, ...this.getInstallArgs()], 'inherit');
   }
 
   protected runRemoveDeps(dependencies: string[]): void {
     const args = ['--ignore-workspace-root-check', ...dependencies];
 
-    this.executeCommand('yarn', ['remove', ...args], 'inherit');
+    this.executeCommand('yarn', ['remove', ...args, ...this.getInstallArgs()], 'inherit');
   }
 
   protected runGetVersions<T extends boolean>(

--- a/code/lib/cli/src/js-package-manager/Yarn1Proxy.ts
+++ b/code/lib/cli/src/js-package-manager/Yarn1Proxy.ts
@@ -8,30 +8,9 @@ export class Yarn1Proxy extends JsPackageManager {
 
   getInstallArgs(): string[] {
     if (!this.installArgs) {
-      this.installArgs = [];
-
-      if (this.detectWorkspaceRoot()) {
-        this.installArgs.push('-W');
-      }
+      this.installArgs = ['--ignore-workspace-root-check'];
     }
     return this.installArgs;
-  }
-
-  detectWorkspaceRoot() {
-    const { workspaces } = this.readPackageJson();
-
-    if (Array.isArray(workspaces)) {
-      return workspaces.length > 0;
-    }
-
-    if (workspaces && typeof workspaces === 'object') {
-      if (workspaces.packages) {
-        return workspaces.packages.length > 0;
-      }
-      return false;
-    }
-
-    return false;
   }
 
   initPackageJson() {
@@ -64,19 +43,19 @@ export class Yarn1Proxy extends JsPackageManager {
   }
 
   protected runAddDeps(dependencies: string[], installAsDevDependencies: boolean): void {
-    let args = ['--ignore-workspace-root-check', ...dependencies];
+    let args = [...dependencies];
 
     if (installAsDevDependencies) {
       args = ['-D', ...args];
     }
 
-    this.executeCommand('yarn', ['add', ...args, ...this.getInstallArgs()], 'inherit');
+    this.executeCommand('yarn', ['add', ...this.getInstallArgs(), ...args], 'inherit');
   }
 
   protected runRemoveDeps(dependencies: string[]): void {
-    const args = ['--ignore-workspace-root-check', ...dependencies];
+    const args = [...dependencies];
 
-    this.executeCommand('yarn', ['remove', ...args, ...this.getInstallArgs()], 'inherit');
+    this.executeCommand('yarn', ['remove', ...this.getInstallArgs(), ...args], 'inherit');
   }
 
   protected runGetVersions<T extends boolean>(

--- a/code/lib/cli/src/js-package-manager/Yarn2Proxy.test.ts
+++ b/code/lib/cli/src/js-package-manager/Yarn2Proxy.test.ts
@@ -27,7 +27,7 @@ describe('Yarn 2 Proxy', () => {
 
       yarn2Proxy.installDependencies();
 
-      expect(executeCommandSpy).toHaveBeenCalledWith('yarn', [], expect.any(String));
+      expect(executeCommandSpy).toHaveBeenCalledWith('yarn', ['install', '-W'], expect.any(String));
     });
   });
 
@@ -69,7 +69,7 @@ describe('Yarn 2 Proxy', () => {
 
       expect(executeCommandSpy).toHaveBeenCalledWith(
         'yarn',
-        ['add', '-D', '@storybook/preview-api'],
+        ['add', '-W', '-D', '@storybook/preview-api'],
         expect.any(String)
       );
     });
@@ -83,7 +83,7 @@ describe('Yarn 2 Proxy', () => {
 
       expect(executeCommandSpy).toHaveBeenCalledWith(
         'yarn',
-        ['remove', '@storybook/preview-api'],
+        ['remove', '-W', '@storybook/preview-api'],
         expect.any(String)
       );
     });

--- a/code/lib/cli/src/js-package-manager/Yarn2Proxy.ts
+++ b/code/lib/cli/src/js-package-manager/Yarn2Proxy.ts
@@ -71,13 +71,13 @@ export class Yarn2Proxy extends JsPackageManager {
       args = ['-D', ...args];
     }
 
-    this.executeCommand('yarn', ['add', ...args, ...this.getInstallArgs()], 'inherit');
+    this.executeCommand('yarn', ['add', ...this.getInstallArgs(), ...args], 'inherit');
   }
 
   protected runRemoveDeps(dependencies: string[]): void {
     const args = [...dependencies];
 
-    this.executeCommand('yarn', ['remove', ...args, ...this.getInstallArgs()], 'inherit');
+    this.executeCommand('yarn', ['remove', ...this.getInstallArgs(), ...args], 'inherit');
   }
 
   protected runGetVersions<T extends boolean>(

--- a/code/lib/cli/src/js-package-manager/Yarn2Proxy.ts
+++ b/code/lib/cli/src/js-package-manager/Yarn2Proxy.ts
@@ -5,6 +5,36 @@ import type { PackageJson } from './PackageJson';
 export class Yarn2Proxy extends JsPackageManager {
   readonly type = 'yarn2';
 
+  installArgs: string[] | undefined;
+
+  getInstallArgs(): string[] {
+    if (!this.installArgs) {
+      this.installArgs = [];
+
+      if (this.detectWorkspaceRoot()) {
+        this.installArgs.push('-W');
+      }
+    }
+    return this.installArgs;
+  }
+
+  detectWorkspaceRoot() {
+    const { workspaces } = this.readPackageJson();
+
+    if (Array.isArray(workspaces)) {
+      return workspaces.length > 0;
+    }
+
+    if (workspaces && typeof workspaces === 'object') {
+      if (workspaces.packages) {
+        return workspaces.packages.length > 0;
+      }
+      return false;
+    }
+
+    return false;
+  }
+
   initPackageJson() {
     return this.executeCommand('yarn', ['init']);
   }
@@ -31,7 +61,7 @@ export class Yarn2Proxy extends JsPackageManager {
   }
 
   protected runInstall(): void {
-    this.executeCommand('yarn', [], 'inherit');
+    this.executeCommand('yarn', ['install', ...this.getInstallArgs()], 'inherit');
   }
 
   protected runAddDeps(dependencies: string[], installAsDevDependencies: boolean): void {
@@ -41,13 +71,13 @@ export class Yarn2Proxy extends JsPackageManager {
       args = ['-D', ...args];
     }
 
-    this.executeCommand('yarn', ['add', ...args], 'inherit');
+    this.executeCommand('yarn', ['add', ...args, ...this.getInstallArgs()], 'inherit');
   }
 
   protected runRemoveDeps(dependencies: string[]): void {
     const args = [...dependencies];
 
-    this.executeCommand('yarn', ['remove', ...args], 'inherit');
+    this.executeCommand('yarn', ['remove', ...args, ...this.getInstallArgs()], 'inherit');
   }
 
   protected runGetVersions<T extends boolean>(

